pubsub/iterator.go

```diff
@@ -490,30 +531,57 @@ func (it *messageIterator) sendAck(m map[string]*AckResult) {
 	for len(ackIDs) > 0 {
 		toSend, ackIDs = splitRequestIDs(ackIDs, ackIDBatchSize)
 
-		recordStat(it.ctx, AckCount, int64(len(toSend)))
-		addAcks(toSend)
-		// Use context.Background() as the call's context, not it.ctx. We don't
-		// want to cancel this RPC when the iterator is stopped.
-		cctx2, cancel2 := context.WithTimeout(context.Background(), 60*time.Second)
-		defer cancel2()
-		err := it.subc.Acknowledge(cctx2, &pb.AcknowledgeRequest{
-			Subscription: it.subName,
-			AckIds:       toSend,
-		})
-		if exactlyOnceDelivery {
-			resultsByAckID := make(map[string]*AckResult)
-			for _, ackID := range toSend {
-				resultsByAckID[ackID] = m[ackID]
+		// Use of anonymous local function allows span.End to be deferred to end of each loop.
+		func() {
+			numBatch := len(toSend)
+			var links []trace.Link
+			if it.enableTracing {
+				for _, ackID := range toSend {
+					// get the parent span context for this ackID for otel tracing.
+					s, _ := it.activeSpans.Load(ackID)
+					parentSpan := s.(trace.Span)
+					defer parentSpan.End()
+					defer parentSpan.SetAttributes(attribute.String(resultAttribute, resultAcked))
+					parentSpan.AddEvent(eventAckStart, trace.WithAttributes(semconv.MessagingBatchMessageCount(numBatch)))
+					defer parentSpan.AddEvent(eventAckEnd)
+					if parentSpan.SpanContext().IsSampled() {
+						links = append(links, trace.Link{SpanContext: parentSpan.SpanContext()})
+					}
+				}
 			}
-			st, md := extractMetadata(err)
-			_, toRetry := processResults(st, resultsByAckID, md)
-			if len(toRetry) > 0 {
-				// Retry acks in a separate goroutine.
-				go func() {
-					it.retryAcks(toRetry)
-				}()
+			ctx := context.Background()
+			var ackSpan trace.Span
+			if it.enableTracing {
+				ctx, ackSpan = startSpan(context.Background(), ackSpanName, it.subID, trace.WithLinks(links...))
+				defer ackSpan.End()
+				ackSpan.SetAttributes(semconv.MessagingBatchMessageCount(numBatch),
```

Comment:
Can you add `semconv.MessagingSystemKey.String("pubsub")` and `semconv.MessagingDestinationName(it.subID)` here and to the other spots?